// @layer0/core/test-utils/MockRequest.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
// @ts-nocheck
const querystring_1 = __importDefault(require("querystring"));
class MockRequest {
    constructor(config) {
        this.method = 'GET';
        this.headers = {
            host: 'jest.layer0.co',
        };
        this.connection = {
            remoteAddress: '127.0.0.1',
        };
        this.setHeader = jest.fn();
        this.updateHeader = jest.fn();
        this.removeHeader = jest.fn();
        this.on = jest.fn((event, cb) => {
            if (event === 'end') {
                cb();
            }
        });
        Object.assign(this, config);
        if (config.url) {
            const [path, qs] = config.url.split(/\?/);
            this.path = path;
            this.query = {
                ...this.query,
                ...(qs && querystring_1.default.parse(qs)),
            };
        }
    }
}
exports.default = MockRequest;
