// @layer0/cli/utils/packageManager.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isYarnInstalled = isYarnInstalled;
exports.isYarn = isYarn;
exports.isYarnVersion = isYarnVersion;
exports.getInstallCommand = getInstallCommand;
exports.runNodePackageManagerScript = runNodePackageManagerScript;
exports.getNodePackageInfo = getNodePackageInfo;
exports.installDependencies = installDependencies;
exports.globalCLIInstallCommand = globalCLIInstallCommand;
exports.installCLIGlobally = installCLIGlobally;
exports.isCLIInstalledGlobally = isCLIInstalledGlobally;

var _ora = _interopRequireDefault(require("ora"));

var _chalk = _interopRequireDefault(require("chalk"));

var _path = require("path");

var _fs = require("fs");

var _run = _interopRequireDefault(require("./run"));

var _logo = _interopRequireDefault(require("./logo"));

var _resolveGlobal = _interopRequireDefault(require("resolve-global"));

var _commandExists = _interopRequireDefault(require("command-exists"));

var _checkNodeVersion = _interopRequireDefault(require("check-node-version"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Returns true if yarn is installed, otherwise false
 * @returns
 */
async function isYarnInstalled() {
  try {
    return await (0, _commandExists.default)('yarn');
  } catch (e) {
    return false;
  }
}
/**
 * Returns if the user is using yarn or npm based on the lock file
 */


function isYarn() {
  return process.env.YARN === 'true' || (0, _fs.existsSync)((0, _path.join)(process.cwd(), 'yarn.lock'));
}
/**
 * Returns `true` if the current yarn version matches the supplied semver
 * @param {string} version
 * @returns
 */


async function isYarnVersion(version) {
  return new Promise(resolve => (0, _checkNodeVersion.default)({
    yarn: String(version)
  }, (error, result) => {
    resolve(result.isSatisfied);
  }));
}
/**
 * Gets the install command for the package manager being used in the user's project
 */


function getInstallCommand() {
  return isYarn() ? 'yarn add --dev' : 'npm i -D';
}
/**
 * Runs npm or yarn command with args based on the package manager used
 * @param {String[]} commandArgs
 */


function runNodePackageManagerScript(commandArgs = [], {
  stdio
} = {}) {
  const command = isYarn() ? 'yarn' : 'npm';
  return (0, _run.default)(command, commandArgs, {
    stdio
  });
}
/**
 *
 * @param {string} packageName Npm package name
 * @returns Package info equal to: npm info @layer0/cli --json
 */


function getNodePackageInfo(packageName) {
  return runNodePackageManagerScript(['info', packageName, '--json']);
}
/**
 * Installs packages if they are not already installed.
 * @param {Object} libs The libs to install.  Keys are the names of node packages and values are versions
 * @param {Object} options
 * @param {Object} options.dev If true, the libs will be installed as devDependencies.
 * @param {Object} options.skipLayer0Deps If true, @layer0 libs will not be installed. Useful for yalc-ing during development
 * @return {Promise}
 */


async function installDependencies(libs, {
  dev = false,
  skipLayer0Deps = false
} = {}) {
  const toInstall = [],
        packageNames = [],
        commandArgs = [];

  for (let lib in libs) {
    if (skipLayer0Deps && lib.startsWith('@layer0')) {
      continue;
    }

    packageNames.push(lib);
    toInstall.push(`${lib}@${libs[lib]}`);
  }

  if (toInstall.length === 0) return Promise.resolve();
  const message = `installing ${_chalk.default.green(packageNames.join(', '))}...`;
  const spinner = (0, _ora.default)(message).start();

  if (isYarn()) {
    commandArgs.push('add');

    if (await isYarnVersion(1)) {
      commandArgs.push('--ignore-workspace-root-check');
    }

    dev && commandArgs.push('--dev');
  } else {
    commandArgs.push('install');
    commandArgs.push(dev ? '--save-dev' : '--save');
  }

  try {
    await runNodePackageManagerScript([...commandArgs, ...toInstall]);
    spinner.succeed(`${message} done.`);
    return Promise.resolve();
  } catch (e) {
    spinner.fail(`${message} failed.\n`);
    return Promise.reject(e);
  }
}
/**
 * Returns the command that the user can run to install @layer0/cli globally using the current package manager.
 * @returns
 */


function globalCLIInstallCommand() {
  if (isYarn()) {
    return 'yarn global add @layer0/cli';
  } else {
    return 'npm i -g @layer0/cli';
  }
}
/**
 * Installs @layer0/cli globally using the current package manager.
 */


async function installCLIGlobally() {
  const message = `installing the ${_logo.default} CLI...`;
  const spinner = (0, _ora.default)(message).start();

  try {
    if (isYarn()) {
      await (0, _run.default)('yarn', ['global', 'add', '@layer0/cli']);
    } else {
      await (0, _run.default)('npm', ['i', '-g', '@layer0/cli']);
    }

    spinner.succeed(`${message} done.`);
  } catch (e) {
    spinner.fail(`${message} failed.\n`);
    throw e;
  }
}
/**
 * Returns true if @layer0/cli is installed globally, otherwise false.
 */


function isCLIInstalledGlobally() {
  try {
    (0, _resolveGlobal.default)('@layer0/cli');
    return true;
  } catch (e) {
    return false;
  }
}
