// @layer0/cli/commands/run.js
"use strict";

const {
  join
} = require('path');

const {
  existsSync
} = require('fs');

const chalk = require('chalk');

const logo = require('../utils/logo');

const runDeploymentArchive = require('../utils/runDeploymentArchive');

const runWithServerless = require('../utils/runWithServerless');

const clearPorts = require('../utils/clearPorts');

const getEntryPoint = require('../utils/getEntryPoint');

const {
  isMonorepo,
  locateAppToRunCmd
} = require('../utils/monorepo');

exports.command = 'run [archive]';
exports.describe = 'Runs your project locally, simulating Layer0 cloud environment. When no arguments are provided, this command is the same as layer0 dev.';

exports.builder = yargs => yargs.option('production', {
  type: 'boolean',
  alias: 'p',
  describe: 'Runs your app in production mode, with caching enabled, emulating Layer0 serverless runtime environment. You must first run layer0 build to create a production build of your app.'
}).option('cache', {
  type: 'boolean',
  alias: 'c',
  describe: 'Enables caching in development mode. Caching is enabled by default in production mode.'
}).positional('archive', {
  describe: 'The path to a deployment archive (zip) file downloaded from Layer0 Developer Console'
});

exports.handler = async ({
  context,
  production,
  archive,
  cache
}) => {
  await clearPorts();

  if (isMonorepo()) {
    await locateAppToRunCmd('run');
  }

  const layer0Dir = join(process.cwd(), '.layer0');
  const {
    logger
  } = context;

  if (process.env.NODE_ENV === 'production') {
    production = true;
    cache = true;
  }

  if (archive) {
    logger.info(`> Running deployed app locally using ${logo}...`);
    await runDeploymentArchive(archive);
  } else if (production) {
    if (!existsSync(layer0Dir)) {
      logger.error(`\nLayer0 production build not found. Please run ${chalk.green('layer0 build')} before running ${chalk.green('layer0 run --production')}.\n`);
      process.exit(1);
    }

    logger.info(`> Running ${logo} in production mode...`); // Layer0 automatically sets this as an environment variable on the Lambda in the cloud

    process.env.NODE_ENV = 'production';
    await runWithServerless(layer0Dir);
  } else {
    logger.info(`> Starting ${logo} in development mode with caching ${cache ? chalk.green('enabled') : chalk.red('disabled')}...`);

    if (cache) {
      process.env.LAYER0_CACHE = 'true';
    }

    const dev = await getEntryPoint('dev');
    dev();
  }
};
