// @layer0/cli/commands/use.js
"use strict";

const Layer0PackageJson = require('../utils/Layer0PackageJson');

const logo = require('../utils/logo');

const {
  installDependencies
} = require('../utils/packageManager');

exports.command = 'use [layer0Version]';
exports.describe = 'Updates Layer0 version';
exports.builder = {
  layer0Version: {
    type: 'string',
    describe: 'Updates Layer0 version',
    demandOption: true
  },
  path: {
    type: 'string',
    describe: "Path to your site's root director. Uses current directory by default",
    default: '.'
  }
};

exports.handler = async ({
  context,
  layer0Version,
  path: projectPath
}) => {
  const packageJson = new Layer0PackageJson(projectPath);
  await context.logger.step(`Updating ${logo}`, async () => {
    context.logger.info(`> Current: ${packageJson.findCurrentLayer0Version()}`);
    context.logger.info(`> New: ${layer0Version}`);
    const devDependencies = Object.fromEntries(packageJson.layer0DevDependencies().map(([name]) => [name, layer0Version]));
    const runtimeDependencies = Object.fromEntries(packageJson.layer0RuntimeDependencies().map(([name]) => [name, layer0Version]));
    await installDependencies(runtimeDependencies, {
      dev: false,
      overrideInstalled: true
    });
    await installDependencies(devDependencies, {
      dev: true,
      overrideInstalled: true
    });
  });
};
