// @layer0/core/router/addBuiltInRoutes.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const cacheManifestFromRouter_1 = __importDefault(require("./cacheManifestFromRouter"));
const environment_1 = require("../environment");
function addBuiltInRoutes(router) {
    addCacheManifest(router);
    try {
        // Add @layer0/devtools built-in routes if the module is installed
        require('@layer0/devtools/addBuiltInRoutes')({
            router,
            local: environment_1.isLocal(),
        });
    }
    catch (e) {
        // istanbul ignore if
        if (e.code !== 'MODULE_NOT_FOUND') {
            throw e;
        }
    }
}
exports.default = addBuiltInRoutes;
/**
 * Adds /__layer0__/cache-manifest.js, which contains:
 *
 * ```
 *  self.__LAYER0_CACHE_MANIFEST__ = {
 *    "/some/route": {
 *      browser: {
 *        serviceWorkerSeconds: (number)
 *      },
 *      edge: {
 *        maxAgeSeconds: (number)
 *      }
 *    }
 *  }
 * ```
 *
 * @private
 * @param router
 */
function addCacheManifest(router) {
    router.match('/__layer0__/cache-manifest.js', ({ cache, setResponseHeader, send, allowCors }) => {
        cache({
            edge: {
                maxAgeSeconds: 60 * 60 * 24 * 365,
            },
            browser: {
                maxAgeSeconds: 60 * 60,
            },
        });
        setResponseHeader('content-type', 'text/javascript');
        allowCors();
        send(() => `self.__LAYER0_CACHE_MANIFEST__=${cacheManifestFromRouter_1.default(router).toJSON()}`);
    });
}
