// @layer0/core/router/validateWriter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * List of header regexes that will be matched against user provided header name.
 * The following headers are not allowed for user to modify in any way.
 *
 * * `x-0-platform` headers are generated and maintained by Layer0 platform to store intermediate state during request processing
 * * `x-0-version` header should only be manipulated by Layer0 platform
 * * `x-0-t` header should only be manipulated by Layer0 platform
 * * `x-0-components` header should only be manipulated by Layer0 platform
 * * `x-0-status` header should only be manipulated by Layer0 platform
 * * `x-0-lambda-name`  header should only be manipulated by Layer0 platform as otherwise it messes with reverse proxying. Removed by XBP
 * * `x-0-lambda-schedule-timeout` header should only be manipulated by Layer0 platform as otherwise it messes with reverse proxying. Removed by XBP
 * * `x-0-received-at` header should only be manipulated by Layer0 platform as otherwise it messes with reverse proxying.
 * * `host` header should only be manipulated by Layer0 platform as otherwise it messes with reverse proxying
 * * `x-request-id` header is generated by Layer0 platform and used as unique identifier of requests
 * * `content-length` header cannot be manipulated apart from manipulating the body and we don't support that in the edge
 * * `via` header is used to prevent request loops in the platform
 */
const LAYER0_PROHIBITED_HEADER_NAMES = [
    // Only allowed for platform to manipulate
    /^\s*x-0-platform\s*/i,
    // Deny list for x-0-* headers
    /^\s*x-0-version\s*$/i,
    /^\s*x-0-t\s*$/i,
    /^\s*x-0-components\s*$/i,
    /^\s*x-0-status\s*$/i,
    /^\s*x-0-lambda-name\s*$/i,
    /^\s*x-0-lambda-schedule-timeout\s*$/i,
    /^\s*x-0-received-at\s*$/i,
    // Deny list for non x-0-* headers
    /^\s*host\s*$/i,
    /^\s*x-request-id\s*$/i,
    /^\s*content-length\s*$/i,
    /^\s*via\s*$/i,
];
/**
 * List of cookie regexes that will be matched against user provided cookie name.
 * The following cookies are not allowed for user to modify in any way.
 *
 * * `layer0_*` cookies are created and maintained by Layer0 platform (e.g. `layer0_buckets` and `layer0_destination`)
 * * `layer0_devtools_*` cookies allow enabling/disabling the devtools on a per-user or per-environment bases
 *    (cf. packages/devtools/src/addBuiltInRoutes.js)
 */
const LAYER0_PROHIBITED_COOKIE_NAMES = [/^layer0_/i];
const LAYER0_WHITELISTED_COOKIE_NAMES = [/^layer0_devtools/i];
/**
 * Validate that prohibited header names are not overrideable by the user
 * @param name
 */
const validateProhibitedHeaderNames = (name) => {
    if (LAYER0_PROHIBITED_HEADER_NAMES.find(prohibitedRegex => name.match(prohibitedRegex))) {
        throw new Error(`Header "${name}" is reserved and cannot be modified`);
    }
};
/**
 * Validate that prohibited cookie names are not overrideable by the user
 * @param name
 */
const validateProhibitedCookieNames = (name) => {
    if (LAYER0_PROHIBITED_COOKIE_NAMES.find(prohibitedRegex => name.match(prohibitedRegex)) &&
        !LAYER0_WHITELISTED_COOKIE_NAMES.find(whitelistRegexp => name.match(whitelistRegexp))) {
        throw new Error(`Cookie "${name}" is reserved and cannot be modified`);
    }
};
/**
 * Validate Set response header
 * @param name header name
 * @param value header value
 */
exports.validateSetResponseHeader = (name /* value */) => {
    validateProhibitedHeaderNames(name);
};
/**
 * Validate Update response header
 * @param name header name
 * @param match match pattern
 * @param replace replace value
 */
exports.validateUpdateResponseHeader = (name /* match, replace */) => {
    validateProhibitedHeaderNames(name);
};
/**
 * Validate Remove response header
 * @param name
 */
exports.validateRemoveResponseHeader = (name) => {
    validateProhibitedHeaderNames(name);
};
/**
 * Validate set request header
 * @param name header name
 * @param value header value
 */
exports.validateSetRequestHeader = (name) => {
    validateProhibitedHeaderNames(name);
};
/**
 * Validate Update request header
 * @param name header name
 * @param match match pattern
 * @param replace replace value
 */
exports.validateUpdateRequestHeader = (name) => {
    validateProhibitedHeaderNames(name);
};
/**
 * Validate Remove request header
 * @param name
 */
exports.validateRemoveRequestHeader = (name) => {
    validateProhibitedHeaderNames(name);
};
/**
 * Validate cookie name.
 * @param name
 */
exports.validateCookieName = (name) => {
    validateProhibitedCookieNames(name);
};
