// @layer0/cli/frameworks/frameworks.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
// istanbul ignore file
// Note: order of keys are important. If we have more
// specific version of the same key, then specific version
// of the same key must go before the less specific version
// for example:
// if we have:
//
// {
//    key: "nuxt",
//    name: "Nuxt.js"
//    builder: "@layer0\nuxt-nitro",
//    framework: 'nuxt',
//    frameworkVersion: "3.x"  <-- specific version must be placed before non specific
//    ...
// }
// {
//    key: "nuxt",
//    name: "Nuxt.js"
//    builder: "@layer0\nuxt",
//    framework: 'nuxt',
//    ...
// }
//
// The versions are compared using "satisfies" function from https://www.npmjs.com/package/semver
//
var _default = [{
  key: 'next',
  name: 'Next.js',
  builder: '@layer0/next',
  devDependencies: ['@layer0/react'],
  framework: 'next',
  frameworkVersion: undefined
}, {
  key: 'spartacus',
  name: 'Spartacus',
  builder: '@layer0/spartacus',
  devDependencies: ['@layer0/angular', 'workbox-build', 'webpack-cli'],
  framework: '@spartacus/core',
  frameworkVersion: undefined
}, {
  key: 'angular',
  name: 'Angular',
  builder: '@layer0/angular',
  devDependencies: ['globby@11.0.2'],
  // later Angular versions use globby@12.x which breaks build
  framework: '@angular/core',
  frameworkVersion: undefined
}, {
  key: 'vsf',
  name: 'Vue Storefront',
  builder: '@layer0/vue-storefront',
  devDependencies: ['@layer0/apollo', '@layer0/vue', '@layer0/nuxt'],
  framework: '@vue-storefront/nuxt',
  frameworkVersion: undefined
}, {
  key: 'nuxt',
  name: 'Nuxt',
  builder: '@layer0/nuxt-nitro',
  devDependencies: ['@layer0/vue'],
  framework: 'nuxt',
  frameworkVersion: '>=3.0.0'
}, {
  key: 'nuxt-v2',
  name: 'Nuxt v2',
  builder: '@layer0/nuxt',
  devDependencies: ['@layer0/vue'],
  framework: 'nuxt',
  frameworkVersion: undefined
}, {
  key: 'gatsby',
  name: 'Gatsby',
  builder: '@layer0/gatsby',
  devDependencies: ['@layer0/react'],
  framework: 'gatsby',
  frameworkVersion: undefined
}, {
  key: 'sapper',
  name: 'Sapper',
  builder: '@layer0/sapper',
  devDependencies: ['@layer0/svelte'],
  framework: 'sapper',
  frameworkVersion: undefined
}, {
  key: 'frontity',
  name: 'Frontity',
  builder: '@layer0/frontity',
  devDependencies: ['@layer0/react'],
  framework: 'frontity',
  frameworkVersion: undefined
}, {
  key: 'fastboot',
  name: 'Ember Fastboot',
  builder: '@layer0/fastboot',
  dependencies: ['@layer0/fastboot'],
  framework: 'ember-cli-fastboot',
  frameworkVersion: undefined
}, {
  key: 'sveltekit',
  name: 'SvelteKit',
  builder: '@layer0/sveltekit',
  dependencies: ['@layer0/sveltekit'],
  devDependencies: ['@layer0/svelte'],
  framework: '@sveltejs/kit',
  frameworkVersion: undefined
}, {
  key: 'razzle',
  name: 'razzle',
  builder: '@layer0/razzle',
  dependencies: ['@layer0/razzle'],
  framework: 'razzle',
  frameworkVersion: undefined
}, {
  key: 'express',
  name: 'express',
  builder: '@layer0/express',
  dependencies: ['@layer0/express'],
  framework: 'express',
  frameworkVersion: undefined
}, {
  key: 'redwoodjs',
  name: 'redwoodjs',
  builder: '@layer0/redwood',
  devDependencies: ['@layer0/redwood'],
  framework: '@redwoodjs/core',
  frameworkVersion: undefined
}, {
  key: 'shopify-hydrogen',
  name: 'shopify-hydrogen',
  builder: '@layer0/shopify-hydrogen',
  devDependencies: ['@layer0/shopify-hydrogen'],
  framework: '@shopify/hydrogen',
  frameworkVersion: undefined
}];
exports.default = _default;
module.exports = exports.default;
