// @layer0/core/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * @typedef {number} JwtAlgo
 *
 * JwtAlgos are specified in the "JSON Web Algorithms" RFC #7518
 *
 * @enum
 */
var JwtAlgo;
(function (JwtAlgo) {
    /** HMAC using SHA-256, symmetric */
    JwtAlgo["HS256"] = "HS256";
    /** HMAC using SHA-384, symmetric */
    JwtAlgo["HS384"] = "HS384";
    /** HMAC using SHA-512, symmetric */
    JwtAlgo["HS512"] = "HS512";
    /** RSASSA-PKCS1-v1_5 using SHA-256, asymmetric, RSA */
    JwtAlgo["RS256"] = "RS256";
    /** RSASSA-PKCS1-v1_5 using SHA-384, asymmetric, RSA */
    JwtAlgo["RS384"] = "RS384";
    /** RSASSA-PKCS1-v1_5 using SHA-512, asymmetric, RSA */
    JwtAlgo["RS512"] = "RS512";
    /** ECDSA using P-256 and SHA-256, asymmetric, ECDSA */
    JwtAlgo["ES256"] = "ES256";
    /** ECDSA using P-384 and SHA-384, asymmetric, ECDSA */
    JwtAlgo["ES384"] = "ES384";
    /** ECDSA using P-512 and SHA-512, asymmetric, ECDSA */
    JwtAlgo["ES512"] = "ES512";
})(JwtAlgo = exports.JwtAlgo || (exports.JwtAlgo = {}));
