// @layer0/core/deploy/paths.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = require("path");
/**
 * The .layer0 build directory
 */
exports.LAYER0_DIR = '.layer0';
/**
 * The destination directory for JS server code
 */
exports.JS_DIR = path_1.join(exports.LAYER0_DIR, 'lambda');
/**
 * The destination directory for static assets
 */
exports.ASSETS_DIR = path_1.join(exports.LAYER0_DIR, 's3');
exports.PERMANENT_ASSETS_DIR = path_1.join(exports.LAYER0_DIR, 's3-permanent');
/**
 * The directory into which source code is copied for debugging purposes
 */
exports.SOURCES_DIR = path_1.join(exports.LAYER0_DIR, 'src');
/**
 * The name of the compiled routes file
 */
exports.ROUTES_FILE_NAME = 'routes.js';
/**
 * Returns the destination path within the lambda dir for the specified backend
 * @param backend A backend name
 * @return a file path
 */
function pathForBackend(backend) {
    return path_1.join('__backends__', `${backend}.js`);
}
exports.pathForBackend = pathForBackend;
